### PR TITLE
En 6209 shuffle with leaving nodes does not remove them

### PR DESCRIPTION
**Pull request by raduchis**
the nodes that get with the rating below the ratingThreshold should be removed from eligible and waiting lists if there still remain enough validators in the shards. After the node has been successfully shuffled out it will be set as "inactive" 

- shuffler removes leaving validators based on PK
- added inactive List for already left validators 
- added unit tests

### CODE REVIEW

File: sharding/xorValidatorShuffler.go [hunk 1]
```diff
@@ -92,23 +93,26 @@ func (rxs *randXORShuffler) UpdateNodeLists(args ArgsUpdateNodes) (map[uint32][]
...
-	}
-
-	// TODO: remove the conditional and the else branch when reinitialization of node in new shard is implemented
-	if rxs.shuffleBetweenShards {
-		return shuffleNodesInterShards(eligibleAfterReshard, waitingAfterReshard, leavingNodes, args.NewNodes, args.Rand)
+		waitingNodes[shard] = vList
 	}
-
-	return shuffleNodesIntraShards(eligibleAfterReshard, waitingAfterReshard, leavingNodes, args.NewNodes, args.Rand)
+	return waitingNodes, leavingNodes
```
Comment (sasurobert) [inline sharding/xorValidatorShuffler.go:115]: add empty line before

File: sharding/xorValidatorShuffler.go [hunk 2]
```diff
@@ -128,17 +132,24 @@ func shuffleNodesInterShards(
 	var shuffledOutNodes []Validator
 	newNbShards := uint32(len(eligible))
 
-	shuffledOutNodes, eligible, leaving = shuffleOutNodes(
+	stillRemainingInLeaving := make([]Validator, 0)
+	stillRemainingInLeaving = append(leaving, stillRemainingInLeaving...)
```
Comment (sasurobert) [inline sharding/xorValidatorShuffler.go:136]: make([]validator, 0, len(leaving))

File: sharding/xorValidatorShuffler.go [hunk 3]
```diff
@@ -152,12 +163,16 @@ func shuffleNodesIntraShards(
 ) (map[uint32][]Validator, map[uint32][]Validator, []Validator) {
 	shuffledOutMap := make(map[uint32][]Validator)
```
Comment (sasurobert) [inline sharding/xorValidatorShuffler.go:164]: add empty line before

File: sharding/xorValidatorShuffler.go [hunk 4]
```diff
@@ -152,12 +163,16 @@ func shuffleNodesIntraShards(
...
 	newNbShards := uint32(len(eligible))
+	stillRemainingInLeaving := make([]Validator, 0)
+	stillRemainingInLeaving = append(leaving, stillRemainingInLeaving...)
```
Comment (sasurobert) [inline sharding/xorValidatorShuffler.go:167]: make([]Validator, 0, len(leaving))

File: sharding/xorValidatorShuffler.go [hunk 5]
```diff
@@ -245,13 +262,13 @@ func shuffleOutShard(
...
 	validators, removed = removeValidatorsFromList(validators, leaving, validatorsToSelect)
-	leaving, _ = removeValidatorsFromList(leaving, removed, len(removed))
+	remainingInLeaving, _ := removeValidatorsFromList(leaving, removed, len(removed))
 
 	validatorsToSelect -= len(removed)
 	shardShuffledEligible := shuffleList(validators, randomness)
 	shardShuffledOut := shardShuffledEligible[:validatorsToSelect]
 	remainingEligible := shardShuffledEligible[validatorsToSelect:]
-	return shardShuffledOut, remainingEligible, leaving
+	return shardShuffledOut, remainingEligible, remainingInLeaving
```
Comment (sasurobert) [inline sharding/xorValidatorShuffler.go:271]: add empty line before

File: sharding/indexHashedNodesCoordinatorWithRater.go [hunk 6]
```diff
@@ -57,6 +57,11 @@ func (ihgs *indexHashedNodesCoordinatorWithRater) ComputeLeaving(allValidators [
 	leavingList := make([]Validator, 0)
 	minChances := ihgs.GetChance(0)
 	for _, vInfo := range allValidators {
+		if vInfo.List == string(core.InactiveList) {
+			log.Debug("inactive validator", "pk", vInfo.GetPublicKey())
```
[the remark this answers is not in the document]
Comment (raduchis) [reply sharding/indexHashedNodesCoordinatorWithRater.go:61]: it is Debug just like the other lists: eligible, waiting, leaving. We could change it in the future.

File: sharding/xorValidatorShuffler.go (hunk 1 above)
Comment (raduchis) [reply sharding/xorValidatorShuffler.go:115]: fixed

File: sharding/xorValidatorShuffler.go (hunk 2 above)
Comment (raduchis) [reply sharding/xorValidatorShuffler.go:136]: fixed

File: sharding/xorValidatorShuffler.go (hunk 3 above)
Comment (raduchis) [reply sharding/xorValidatorShuffler.go:164]: fixed

File: sharding/xorValidatorShuffler.go (hunk 4 above)
Comment (raduchis) [reply sharding/xorValidatorShuffler.go:167]: fixed

File: sharding/xorValidatorShuffler.go (hunk 5 above)
Comment (raduchis) [reply sharding/xorValidatorShuffler.go:271]: fixed

**Pull request merged**